xdsl/ir/core.py

```diff
@@ -769,30 +769,38 @@ def verify(self, verify_nested_ops: bool = True) -> None:
             if isinstance(operand, ErasedSSAValue):
                 raise Exception("Erased SSA value is used by the operation")
 
-        if (parent_block := self.parent) is not None and (
-            parent_region := parent_block.parent
-        ) is not None:
-            if self.successors and parent_block.last_op != self:
+        parent_block = self.parent
+        parent_region = None if parent_block is None else parent_block.parent
+
+        if self.successors:
+            if parent_block is None or parent_region is None:
                 raise VerifyException(
-                    "Operation with block successors must terminate its parent block"
+                    f"Operation {self.name} with block successors does not belong to a block or a region"
+                )
+
+            if parent_block.last_op is not self:
+                raise VerifyException(
+                    f"Operation {self.name} with block successors must terminate its parent block"
                 )
 
             for succ in self.successors:
-                if succ.parent != parent_region:
-                    raise VerifyException("Branching to a block of a different region")
+                if succ.parent != parent_block.parent:
+                    raise VerifyException(
+                        f"Operation {self.name} is branching to a block of a different region"
+                    )
 
+        if parent_block is not None and parent_region is not None:
             # TODO single-block regions dealt when the NoTerminator trait is
             # implemented (https://github.com/xdslproject/xdsl/issues/1093)
             if len(parent_region.blocks) > 1:
-                if parent_block.last_op is self and not self.has_trait(IsTerminator):
+                if (
+                    parent_block.last_op is self
+                    and len(self.successors) == 0
```

Comment:
```suggestion
```
Even if an operation has successor, it should be a terminator?

Reply:
Yeah, that's a good catch, I re-read the MLIR `Verifier.cpp` wrong on this.